python/pythonmonkey/js_apis/console.js

```diff
@@ -0,0 +1,919 @@
+// @ts-check
+///<reference path="./global.d.ts"/>
+
+const { defineGlobal } = internalBinding("utils")
+const {
+  isAnyArrayBuffer,
+  isPromise,
+  isRegExp,
+  isTypedArray,
+  getPromiseDetails,
+  getProxyDetails,
+} = internalBinding("utils")
+
+const customInspectSymbol = Symbol.for('nodejs.util.inspect.custom')
+
+// Keep this in sync with both SpiderMonkey and V8
+/** @type {PromiseState.Pending} */
+const kPending = 0;
+/** @type {PromiseState.Fulfilled} */
+const kFulfilled = 1;
+/** @type {PromiseState.Rejected} */
+const kRejected = 2;
+
+/**
+ * @param {string[]} output 
+ * @param {string} separator 
+ */
+function join(output, separator) {
+  return output.join(separator)
+}
+
+/**
+ * @return {o is Error}
+ */
+function isError(e) {
+  return objectToString(e) === '[object Error]' || e instanceof Error;
+}
+
+/**
+ * @return {o is Date}
+ */
+function isDate(o) {
+  return objectToString(o) === '[object Date]' || o instanceof Date;
+}
+
+/**
+ * @return {o is Set}
+ */
+function isSet(o) {
+  return objectToString(o) === '[object Set]' || o instanceof Set;
+}
+
+/**
+ * @return {o is Map}
+ */
+function isMap(o) {
+  return objectToString(o) === '[object Map]' || o instanceof Map;
+}
+
+/**
+ * @return {o is DataView}
+ */
+function isDataView(o) {
+  return objectToString(o) === '[object DataView]' || o instanceof DataView;
+}
+
+/**
+ * V8 IsJSExternalObject
+ */
+// TODO (Tom Tang): What's the equivalent in SpiderMonkey?
+function isExternal(o) {
+  return false;
+}
+
+function objectToString(o) {
+  return Object.prototype.toString.call(o);
+}
```

Comment:
Helper functions for the Node.js code because in Node.js they are mostly implemented in C++.